java/arcs/core/data/CapabilityNew.kt

```diff
@@ -0,0 +1,131 @@
+
+    abstract fun compare(other: CapabilityNew): Comparison
+
+    data class Persistence(val type: Type) : CapabilityNew() {
+        enum class Type { None, InMemory, OnDisk, Unrestricted }
+
+        override fun compare(other: CapabilityNew): Comparison {
+            return when (other) {
+                is Persistence -> {
+                    if (type.ordinal < other.type.ordinal) {
+                        Comparison.Stricter
+                    } else if (type.ordinal > other.type.ordinal) {
+                        Comparison.LessStrict
+                    } else Comparison.Equivalent
+                }
+                else -> throw IllegalArgumentException("Cannot compare Persistence with $other")
+            }
+        }
+
+        companion object {
+            fun unrestricted() = Persistence(Type.Unrestricted)
+            fun onDisk() = Persistence(Type.OnDisk)
+            fun inMemory() = Persistence(Type.InMemory)
+            fun none() = Persistence(Type.None)
+        }
+    }
+
+    sealed class TtlNew(count: Int, val isInfinite: Boolean = false) : CapabilityNew() {
+        val millis: Int = count * when (this) {
+            is Millis -> 1
+            is Minutes -> 1 * MILLIS_IN_MIN
+            is Hours -> 60 * MILLIS_IN_MIN
+            is Days -> 60 * 24 * MILLIS_IN_MIN
+            is Infinite -> 1
+        }
+        init {
+            require(count > 0 || isInfinite) {
+                "must be either positive count on infinite, " +
+                    "but got count=$count and isInfinite=$isInfinite"
+            }
+        }
+
+        override fun compare(other: CapabilityNew): Comparison {
+            return when (other) {
+                is TtlNew -> {
+                    if ((isInfinite && other.isInfinite) || millis == other.millis) {
+                        Comparison.Equivalent
+                    } else if (isInfinite) {
+                        Comparison.LessStrict
+                    } else if (other.isInfinite) {
+                        Comparison.Stricter
+                    } else if (millis < other.millis) {
+                        Comparison.Stricter
+                    } else Comparison.LessStrict
+                }
+                else -> throw IllegalArgumentException("Cannot compare Ttl with $other")
+            }
+        }
+
+        fun calculateExpiration(time: Time): Long =
+            if (isInfinite) RawEntity.UNINITIALIZED_TIMESTAMP
+            else time.currentTimeMillis + millis
+
+        data class Millis(val count: Int) : TtlNew(count)
+        data class Minutes(val count: Int) : TtlNew(count)
+        data class Hours(val count: Int) : TtlNew(count)
+        data class Days(val count: Int) : TtlNew(count)
+        data class Infinite(val count: Int = TTL_INFINITE) : TtlNew(count, true)
+
+        companion object {
+            const val TTL_INFINITE = -1
+            const val MILLIS_IN_MIN = 60 * 1000
```

Comment:
Should this be `Double`?

Reply:
changed to Long, do you think Double is better suited?

Reply:
I think `Long` is fine.